src/rules/import-alias.ts

```diff
@@ -10,23 +10,59 @@ import type { JSONSchema4 } from "json-schema";
 import { dirname, join as joinPath, resolve, sep as pathSep } from "path";
 import slash from "slash";
 
+const isPathImport = (
+    relativeImportOverride: RelativeImportConfig
+): relativeImportOverride is RelativePathConfig => {
+    return (relativeImportOverride as RelativePathConfig).path !== undefined;
+};
+
+const isGlobImport = (
+    relativeImportOverride: RelativeImportConfig
+): relativeImportOverride is RelativeGlobConfig => {
+    return (relativeImportOverride as RelativeGlobConfig).pattern !== undefined;
+};
+
 function isPermittedRelativeImport(
     importModuleName: string,
     relativeImportOverrides: RelativeImportConfig[],
-    fileAbsoluteDir: string
+    fileAbsoluteDir: string,
+    filepath: string
 ) {
-    const configOfPath = relativeImportOverrides.find((config) =>
-        fileAbsoluteDir.includes(resolve(config.path))
-    );
-    if (!configOfPath) {
-        return false;
-    }
+    console.log({
+        importModuleName,
+        relativeImportOverrides,
+        fileAbsoluteDir,
+        filepath,
+    });
+    for (const relativeImportOverride of relativeImportOverrides) {
+        const importParts = importModuleName.split("/");
+        const relativeDepth = importParts.filter(
+            (moduleNamePart) => moduleNamePart === ".."
+        ).length;
+
+        if (isPathImport(relativeImportOverride)) {
+            const configOfPath = fileAbsoluteDir.includes(
+                resolve(relativeImportOverride.path)
+            );
+            if (!configOfPath) {
+                return false;
+            }
+            return relativeImportOverride.depth >= relativeDepth;
+        }
 
-    const importParts = importModuleName.split("/");
-    const relativeDepth = importParts.filter(
-        (moduleNamePart) => moduleNamePart === ".."
-    ).length;
-    return configOfPath.depth >= relativeDepth;
+        if (isGlobImport(relativeImportOverride)) {
```

Comment:
I think TypeScript should be able to type narrow simply from the previous `if` condition. It might be nice to get rid of this condition, so that there was an additional type in the type union, TS would be unable to narrow the type and we'd get a compilation error.

Another alternative could be to return false outside of this condition; however, that feels more error prone than having a compilation failure force an update.

Reply:
to get TS to handle the union, i added never. Anyway you will see

> however, that feels more error prone than having a compilation failure force an update.
i did not understand

Reply:
In your previous iteration, if you simply removed the if condition, it would have been sufficient to trigger a compiler error if we added a new element to the union, which was preferable and got rid of the inconsistent return type (with it being `boolean | undefined` before)

For example, if a new contributor added a `RelativeFooConfig`

```
interface RelativeFooConfig {
    foo: "bar";
    depth: number;
}

type RelativeImportConfig = RelativePathConfig | RelativeGlobConfig | RelativeFooConfig;
```

They would immediately be prompted to fix the compilation error if we didn't include an if statement.

```
function isPermittedRelativeImport(
    importModuleName: string,
    relativeImportOverrides: RelativeImportConfig[],
    fileAbsoluteDir: string,
    filepath: string
) {
    for (const relativeImportOverride of relativeImportOverrides) {
        const importParts = importModuleName.split("/");
        const relativeDepth = importParts.filter(
            (moduleNamePart) => moduleNamePart === ".."
        ).length;

        if (isPathImport(relativeImportOverride)) {
            const configOfPath = fileAbsoluteDir.includes(
                resolve(relativeImportOverride.path)
            );
            if (!configOfPath) {
                return false;
            }
            return relativeImportOverride.depth >= relativeDepth;
        }

        const { pattern, depth } = relativeImportOverride; ■ Property 'pattern' does not exist on type 'RelativeGlobConfig | RelativeFooConfig'.

        if (pattern instanceof RegExp && pattern.test(filepath)) {
            return depth >= relativeDepth;
        }

        const filename = filepath.split("/").at(-1) ?? "";
        if (typeof pattern === "string" && filename === pattern) {
            return depth >= relativeDepth;
        }
    }
}
```